### PR TITLE
v2 – 1st draft for Pocket import via API & Wallabag v1 import

### PR DESCRIPTION
**Pull request by nicosomb**
Issue #236

How to use the Pocket Import:
* Access to http://localhost:8000/import
* Click on "Connect to Pocket"
* Log in to Pocket and authorize the application to access to your account
* Entries are imported. 

TODO 
- [x] we should create an interface or an abstract class to define an import
- [x] create an import for each read it later out there (Readability, Pocket, Instapaper) which extends this class
- [x] each import should be a service to put less stuff in the controler
- [x] [Working with Tagged Services](http://symfony.com/doc/current/components/dependency_injection/tags.html) to list importers in Import::index.html.twig
- [x] write tests that mock each service (wow, big step)
- ~~new step before importing to select which entries we want to import.~~ Not easily doable in a first time (because Pocket only provides to get all entries, impossible to get entry one per one)
- [x] manage API responses (errors, etc.)
- ~~use `since` parameter in Pocket API~~ removed this feature. 

In fact, entries are created with excerpt, title, preview picture, but not content (Pocket limitation, see below). 

Some readings: 
* rate limits: https://getpocket.com/developer/docs/rate-limits
* article view https://getpocket.com/developer/docs/v3/article-view Impossible to get articles content, we need to pay for that. So we need to use graby. And we need to discuss about performance. And maybe use [JMSJobQueueBundle](http://jmsyst.com/bundles/JMSJobQueueBundle/master/installation).

Need a first review to see if I'm doing it wrong.

### CODE REVIEW

**Comment by j0k3r** [comment]
I know it's a first draft, but here are my thoughts:
- we should create an interface or an abstract class to define an import
- create an import for each read it later out there (Readability, Pocket, Instapaper) which extends this class
- each import should be a service to put less stuff in the controler
- write tests that mock each service

Once this part is done, we can think about async insertion in case of content need to be fetched on our side.

For that part, we can use a queue system (redis, rabbit, etc ..):
- the service push new entry in the queue
- the worker grab that entry and fetch the content

The major downside of this method is it won't be available on a shared hosting. Maybe we can just use a different method for that with a _basic_ queue system using a cron (I hope shared hosting give ability to use cron...) with a table in the database which acts as the queue.

**Comment by tcitworld** [comment]
Won't Pocket limit the calls made with the `pocket_consumer_key` at some point ?

**Comment by nicosomb** [comment]
That's what I said to @j0k3r on our gitter. 
* the import is user friendly, but pocket can limit the calls
* the import is not user friendly (each user creates its own app on pocket website), and no limit from pocket.

**Comment by tcitworld** [comment]
I've seen https://getpocket.com/developer/docs/rate-limits and we should be fine for a while.

**Comment by nicosomb** [comment]
I followed @j0k3r's advices, and I did:
> * we should create an interface or an abstract class to define an import
> * create an import for each read it later out there (Readability, Pocket, Instapaper) which extends this class
> * each import should be a service to put less stuff in the controler

Next steps:
- [ ] [Working with Tagged Services](http://symfony.com/doc/current/components/dependency_injection/tags.html) to list importers in Import::index.html.twig
- [ ] write tests that mock each service (wow, big step)
- [ ] new step before importing to select which entries we want to import.
- [ ] manage API responses (errors, etc.)

File: src/Wallabag/ImportBundle/Import/PocketImport.php [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+     */
+    private function guessTitle($pocketEntry)
+    {
+        if (isset($pocketEntry['resolved_title']) && $pocketEntry['resolved_title'] != '') {
+            return $pocketEntry['resolved_title'];
+        } elseif (isset($pocketEntry['given_title']) && $pocketEntry['given_title'] != '') {
+            return $pocketEntry['given_title'];
+        } else {
+            return 'Untitled';
+        }
```
Comment (j0k3r) [inline src/Wallabag/ImportBundle/Import/PocketImport.php]: You don't need that `else`.
Just write the `return`

**Comment by nicosomb** [comment]
For this PR, I need help for this part `write tests that mock each service` :-(

**Comment by j0k3r** [comment]
I can help you when I'll have some time

**Comment by j0k3r** [comment]
@nicosomb I did a test sample, you know have to test all cases (error from Pocket, etc ..) :rocket:

**Comment by nicosomb** [comment]
OK I will have a look

**Comment by j0k3r** [comment]
Import screenshots

**Comment by j0k3r** [comment]
I think we have a pretty good first version for the next alpha. There are still some improvements, mostly related if the import is really huge. And in that case we should plan a queue system.

**Comment by nicosomb** [comment]
About Pocket import, we are in this case ?
> the import is not user friendly (each user creates its own app on pocket website), and no limit from pocket

Or we'll create a global application on pocket website?

**Comment by j0k3r** [comment]
@nicosomb I think we can keep a Pocket application for framabag.
For a manual install of wallabag, user'll have to create its own Pocket application. We should add that to the documentation.

**Comment by nicosomb** [comment]
What will we have to do on this PR?

**Comment by j0k3r** [comment]
Nothing, it's ready for a first version.
We'll need to some async for huge import in the next version.

**Comment by nicosomb** [comment]
I'm testing all this PR and adding french translations.

**Comment by nicosomb** [comment]
Warning during pocket import: 

```
Warning: DOMNode::cloneNode(): ID installation already defined
```

in vendor/j0k3r/graby/src/Extractor/ContentExtractor.php at line 621 `$this->body = $this->body->cloneNode(true);`

**Comment by nicosomb** [comment]
Sorry for the tests fails, my local testsuite is broken due to my environment. I will fix my environment. 

I still have 2 errors (`PHPUnit_Framework_Exception: Argument #2 (No Value) of PHPUnit_Framework_Assert::assertContains() must be a array, traversable or string`), I don't know why, will have a look.

**Comment by j0k3r** [comment]
About the graby error, it seems that the website is down
> It's not just you! http://devblog.lexik.fr looks down from here.

**Comment by nicosomb** [comment]
> About the graby error, it seems that the website is down
Yesterday, it was OK.

**Comment by nicosomb** [comment]
still a problem, URL in description is wrong displayed

**Comment by nicosomb** [comment]
@j0k3r URL is displayed, instead of being a real link. Any idea?

**Comment by j0k3r** [comment]
Do we really need a link to getpocket.com ?

**Comment by nicosomb** [comment]
No, you're right.

File: src/Wallabag/ImportBundle/Resources/views/Import/index.html.twig [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+{% block content %}
+<div class="row">
+    <div class="col s12">
+        <div class="card-panel settings">
+            {% trans %}Welcome on wallabag importer. Please select your previous service that you want to migrate.{% endtrans %}
+            <ul>
+                {% for import in imports %}
+                    <li>
+                        <h5>{{ import.name }}</h5>
+                        <blockquote>{{ import.description|raw|trans }}</blockquote>
```
Comment (j0k3r) [inline src/Wallabag/ImportBundle/Resources/views/Import/index.html.twig]: Do we really need `|raw` since we didn't include any tags now?

**Comment by nicosomb** [comment]
all seems to be ok now.

**Comment by j0k3r** [comment]
:rocket:

**Pull request merged**